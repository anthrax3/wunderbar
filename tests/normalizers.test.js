const data = require('./__fixtures__/age-distrib-germany.json');

const {
  normalizeValues,
  normalizeOptions,
  createNormalizeValue,
} = require('../lib/normalizers');

describe('normalizers', () => {
  describe('normalizeValues', () => {
    it('should throw if values is not an array', () => {
      expect(() => normalizeValues('foo')).toThrowErrorMatchingSnapshot();
    });

    it('should throw if values are not numbers', () => {
      expect(() =>
        normalizeValues(['foo', () => {}])
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if there is no value key in values', () => {
      expect(() =>
        normalizeValues([{ rabbit: 1 }, {}])
      ).toThrowErrorMatchingSnapshot();
    });

    it('should normalize Array<number> to Array<{ value: number }>', () => {
      expect(normalizeValues([1, 2, 3])).toMatchSnapshot();
    });

    it('should normalize Array<string> to Array<{ value: number }>', () => {
      expect(normalizeValues(['1', '2', '3'])).toMatchSnapshot();
    });
  });

  describe('normalizeOptions', () => {
    it('should properly print NaN when in error message', () => {
      expect(() =>
        normalizeOptions({ min: NaN })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if min is not a number', () => {
      expect(() =>
        normalizeOptions({ min: [] })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if max is not a number', () => {
      expect(() =>
        normalizeOptions({ max: {} })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if length is not a number', () => {
      expect(() =>
        normalizeOptions({ length: () => {} })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if sort type is not supported', () => {
      expect(() =>
        normalizeOptions({ sort: 'magic' })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if sort is not a function', () => {
      expect(() =>
        normalizeOptions({ sort: {} })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if format is not a string or a function', () => {
      expect(() =>
        normalizeOptions({ format: null })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if randomColorOptions is not an object', () => {
      expect(() =>
        normalizeOptions({ randomColorOptions: null })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should return default values if options are empty', () => {
      expect(normalizeOptions({})).toMatchSnapshot();
    });

    it('should normalize min, max and length to numbers, if strings are provided', () => {
      expect(
        normalizeOptions({ min: '1.23', max: '13.5', length: '42' })
      ).toMatchSnapshot();
    });
  });

  describe('createNormalizeValue', () => {
    const colors = ['red', 'blue', 'green', 'pink', 'yellow'];
    const valueFormatter = value => `meow ${value.toFixed(2)}`;
    const chartLength = 100;

    const createNormalize = (min, max) =>
      createNormalizeValue(min, max, chartLength, colors, valueFormatter);

    describe('normalizeValue', () => {
      it('should prepare data for drawing', () => {
        const normalized = createNormalize(0, data[0].value);
        expect(normalized(data[0], 0)).toMatchSnapshot();
      });

      it('should cap data value by maxValue', () => {
        const normalized = createNormalize(0, 500);
        expect(normalized(data[0], 0).normalizedValue).toEqual(500);
      });

      it('should shift data value by minValue', () => {
        const normalized = createNormalize(1000, data[0].value);
        expect(normalized(data[0], 0).normalizedValue).toEqual(
          data[0].value - 1000
        );
      });

      it('should calculate lineLength based on chartLength and normalizedValue', () => {
        const normalized = createNormalize(0, data[0].value);

        // Expecting chartLength here because maxValue of chart is the same as data value
        expect(normalized(data[0], 0).lineLength).toEqual(chartLength);
      });

      it('should return formatted value', () => {
        const normalized = createNormalize(0, data[0].value);
        expect(normalized(data[0], 0).formattedValue).toEqual(
          valueFormatter(data[0].value)
        );
      });
    });
  });
});
