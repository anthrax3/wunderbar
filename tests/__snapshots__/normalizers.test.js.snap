// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizers createNormalizeValue normalizeValue should prepare data for drawing 1`] = `
Object {
  "chartBar": "â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ",
  "color": "red",
  "coloredChartBar": "[31mâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ[39m",
  "formattedValue": "meow 10333117.00",
  "label": "0-14",
  "lineLength": 100,
  "normalizedValue": 10333117,
  "rawValue": 10333117,
}
`;

exports[`normalizers normalizeOptions should normalize min, max and length to numbers, if strings are provided 1`] = `
Object {
  "format": "0.00",
  "length": 42,
  "max": 13.5,
  "min": 1.23,
  "randomColorOptions": Object {},
  "sort": "none",
}
`;

exports[`normalizers normalizeOptions should properly print NaN when in error message 1`] = `"Unexpected options.min: NaN. Expected number"`;

exports[`normalizers normalizeOptions should return default values if options are empty 1`] = `
Object {
  "format": "0.00",
  "length": undefined,
  "max": undefined,
  "min": undefined,
  "randomColorOptions": Object {},
  "sort": "none",
}
`;

exports[`normalizers normalizeOptions should throw if format is not a string or a function 1`] = `"Unexpected options.format: null. Expected function or string"`;

exports[`normalizers normalizeOptions should throw if length is not a number 1`] = `"Unexpected options.length: [Function]. Expected number"`;

exports[`normalizers normalizeOptions should throw if max is not a number 1`] = `"Unexpected options.max: {}. Expected number"`;

exports[`normalizers normalizeOptions should throw if min is not a number 1`] = `"Unexpected options.min: []. Expected number"`;

exports[`normalizers normalizeOptions should throw if randomColorOptions is not an object 1`] = `"Unexpected options.randomColorOptions: null. Expected object with randomcolor options (https://github.com/davidmerfield/randomColor#options)"`;

exports[`normalizers normalizeOptions should throw if sort is not a function 1`] = `"Unexpected options.sort: {}. Expected function or \\"min\\" | \\"max\\" | \\"none\\""`;

exports[`normalizers normalizeOptions should throw if sort type is not supported 1`] = `"Unexpected options.sort: \\"magic\\". Expected function or \\"min\\" | \\"max\\" | \\"none\\""`;

exports[`normalizers normalizeValues should normalize Array<number> to Array<{ value: number }> 1`] = `
Array [
  Object {
    "value": 1,
  },
  Object {
    "value": 2,
  },
  Object {
    "value": 3,
  },
]
`;

exports[`normalizers normalizeValues should normalize Array<string> to Array<{ value: number }> 1`] = `
Array [
  Object {
    "value": 1,
  },
  Object {
    "value": 2,
  },
  Object {
    "value": 3,
  },
]
`;

exports[`normalizers normalizeValues should throw if there is no value key in values 1`] = `"Unexpected value { \\"rabbit\\": 1 } in \\"values\\" argument. Expected { value: number, color?: string, label?: string }"`;

exports[`normalizers normalizeValues should throw if values are not numbers 1`] = `"Unexpected value \\"foo\\" in \\"values\\" argument. Expected { value: number, color?: string, label?: string }"`;

exports[`normalizers normalizeValues should throw if values is not an array 1`] = `"Unexpected \\"values\\" argument \\"foo\\". Expected Array<{ value: number, color?: string, label?: string }>"`;
